File: src/androidTest/java/com/todoroo/astrid/subtasks/SubtasksTestCase.java
package com.todoroo.astrid.subtasks;

import com.todoroo.astrid.api.Filter;
import com.todoroo.astrid.core.BuiltInFilterExposer;
import com.todoroo.astrid.dao.TaskListMetadataDao;
import com.todoroo.astrid.data.Task;
import com.todoroo.astrid.data.TaskListMetadata;
import com.todoroo.astrid.service.TaskService;
import com.todoroo.astrid.subtasks.AstridOrderedListUpdater.Node;
import com.todoroo.astrid.test.DatabaseTestCase;

import org.tasks.preferences.Preferences;

import javax.inject.Inject;

/**
 * Contains useful methods common to all subtasks tests
 * @author Sam
 *
 */
public class SubtasksTestCase extends DatabaseTestCase {

    @Inject TaskListMetadataDao taskListMetadataDao;
    @Inject TaskService taskService;
    @Inject Preferences preferences;

    protected SubtasksUpdater<TaskListMetadata> updater;
    protected Filter filter;

    /* Starting State:
    *
    * A
    *  B
    *  C
    *   D
    * E
    * F
    */
    public static final String DEFAULT_SERIALIZED_TREE = "[-1, [1, 2, [3, 4]], 5, 6]".replaceAll("\\s", "");

    @Override
    protected void setUp() {
        super.setUp();
        filter = BuiltInFilterExposer.getMyTasksFilter(getContext().getResources());
        preferences.clear(SubtasksUpdater.ACTIVE_TASKS_ORDER);
        updater = new SubtasksFilterUpdater(taskListMetadataDao, taskService);
    }

    protected void expectParentAndPosition(Task task, Task parent, int positionInParent) {
        String parentId = (parent == null ? "-1" : parent.getUuid());
        Node n = updater.findNodeForTask(task.getUuid());
        assertNotNull("No node found for task " + task.getTitle(), n);
        assertEquals("Parent mismatch", parentId, n.parent.uuid);
        assertEquals("Position mismatch", positionInParent, n.parent.children.indexOf(n));
    }

}

File: app/src/androidTest/java/com/todoroo/astrid/subtasks/SubtasksTestCase.kt
package com.todoroo.astrid.subtasks

import androidx.test.InstrumentationRegistry
import com.todoroo.astrid.core.BuiltInFilterExposer
import com.todoroo.astrid.dao.TaskDao
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.tasks.data.dao.TaskListMetadataDao
import org.tasks.data.entity.Task
import org.tasks.filters.AstridOrderingFilter
import org.tasks.injection.InjectingTestCase
import org.tasks.preferences.Preferences
import javax.inject.Inject

abstract class SubtasksTestCase : InjectingTestCase() {
    lateinit var updater: SubtasksFilterUpdater
    lateinit var filter: AstridOrderingFilter
    @Inject lateinit var taskListMetadataDao: TaskListMetadataDao
    @Inject lateinit var taskDao: TaskDao
    @Inject lateinit var preferences: Preferences
    
    override fun setUp() {
        super.setUp()
        filter = BuiltInFilterExposer.getMyTasksFilter(InstrumentationRegistry.getTargetContext().resources)
        preferences.clear(SubtasksFilterUpdater.ACTIVE_TASKS_ORDER)
        updater = SubtasksFilterUpdater(taskListMetadataDao, taskDao)
    }

    fun expectParentAndPosition(task: Task, parent: Task?, positionInParent: Int) {
        val parentId = parent?.uuid ?: "-1"
        val n = updater.findNodeForTask(task.uuid)
        assertNotNull("No node found for task " + task.title, n)
        assertEquals("Parent mismatch", parentId, n!!.parent!!.uuid)
        assertEquals("Position mismatch", positionInParent, n.parent!!.children.indexOf(n))
    }

    companion object {
        /* Starting State:
   *
   * A
   *  B
   *  C
   *   D
   * E
   * F
   */
        val DEFAULT_SERIALIZED_TREE = "[-1, [1, 2, [3, 4]], 5, 6]".replace("\\s".toRegex(), "")
    }
}